README.md
<div id="top"></div>

# Simba Grid

Infinite grid scroll.

## [View Online Demo](https://rodgath.github.io/simba-grid/demo/)

<br>

[![NPM Version](https://img.shields.io/npm/v/simba-grid.svg?color=00BAB8)]()
[![NPM License](https://img.shields.io/npm/l/all-contributors.svg?color=00a7a6)](https://github.com/Rodgath/simba-grid/blob/master/LICENSE)
[![NPM Downloads](https://img.shields.io/npm/dw/simba-grid.svg?color=009593)]() 
![npm bundle size (version)](https://img.shields.io/bundlephobia/minzip/simba-grid/latest?color=008281)

![GitHub issues](https://img.shields.io/github/issues-raw/Rodgath/simba-grid?color=37bf4c)
[![GitHub pull requests](https://img.shields.io/github/issues-pr/Rodgath/simba-grid.svg?color=32ac44)]()
[![PR's Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?color=278635)](http://makeapullrequest.com) 

# Table of Contents

- [Installation](#installation)
- [Implementation](#implementation)
  - [Method 1](#method-1)
  - [Method 2](#method-2)
  - [Method 3](#method-3)
- [Options](#options)
- [License](#license)

# Installation

**[A]** Download 

+ [simba-grid.min.js](https://unpkg.com/simba-grid@latest/dist/js/simba-grid.min.js) - Minified
+ [simba-grid.js](https://unpkg.com/simba-grid@latest/dist/js/simba-grid.min.js) - Unminified

**[B]** Package

+ Install with [npm](https://www.npmjs.com/package/simba-grid): `npm install simba-grid` 
* Install with [yarn](https://yarnpkg.com/): `yarn add simba-grid`

* [![NPM](https://nodei.co/npm/simba-grid.png?downloads=true)](https://www.npmjs.com/package/simba-grid)

**[C]** Get a local working copy of the development repository _(Optional)_ <br />
`git clone https://github.com/Rodgath/simba-grid.git`


# Implementation
There are several methods you can use to add Simba Grid into your project.

## Method 1

#### Initializing with `simbaGrid` function. [View Demo](https://rodgath.github.io/simba-grid/demo/index.html)
Your HTML code with images
```html
<div class="my-simba-grid">
  <img src="./images/01.jpg" alt="Image 01">
  <img src="./images/02.jpg" alt="Image 02">

  <!-- More Images Here -->
  
  <img src="./images/20.jpg" alt="Image 20">
  <img src="./images/21.jpg" alt="Image 21">
</div>
```

#### Call the `simbaGrid()` function with two arguments.
...**1)** The element _'class'_ or _'id'_ holding the images 

...**2)** The _'options'_ object
```javascript
document.addEventListener('DOMContentLoaded', function () {
  simbaGrid('.my-simba-grid', {
    animationStyle : 'rotate'
  });
});
```

#### Enqueue the simbaGrid script at the bottom of your markup
+ Using local script file.
```html
<script src="simba-grid.min.js"></script>
<!-- OR -->
<script src="./node_modules/simba-grid/dist/js/simba-grid.min.js"></script>
```
+ Using CDN file. _(Optional)_
```html
<script src="https://cdn.jsdelivr.net/npm/simba-grid@latest/dist/js/simba-grid.min.js"></script>
```

##### <div align="right"><a href="#top">&uarr; TOP &uarr;</a></div>

---
## Method 2

#### Using `data-simba-grid` attribute. [View Demo](https://rodgath.github.io/simba-grid/demo/index3.html)
Your HTML code with images
```html
<div data-simba-grid='{ "animationStyle": "zoomRotate", "cols": "4" }'>
  <img src="./images/01.jpg" alt="Image 01">
  <img src="./images/02.jpg" alt="Image 02">

  <!-- More Images Here -->
  
  <img src="./images/20.jpg" alt="Image 20">
  <img src="./images/21.jpg" alt="Image 21">
</div>
```

#### Enqueue the simbaGrid script at the bottom of your markup
+ Using local script file.
```html
<script src="simba-grid.min.js"></script>
<!-- OR -->
<script src="./node_modules/simba-grid/dist/js/simba-grid.min.js"></script>
```
+ Using CDN file. _(Optional)_
```html
<script src="https://cdn.jsdelivr.net/npm/simba-grid@latest/dist/js/simba-grid.min.js"></script>
```

##### <div align="right"><a href="#top">&uarr; TOP &uarr;</a></div>

---
## Method 3

#### Using JSON object of images
...**[A]** - Adding the JSON object inside the `data-simba-grid` attribute. [View Demo](https://rodgath.github.io/simba-grid/demo/index4.html)

```html
<div data-simba-grid='{ "animationStyle": "zoom", 
    "cols": "4", 
    "rows": "4", 
    "gap": "4", 
    "rowHeight": "150", 
    "images": [
      { "src": "./images/01.jpg", "title": "Image 01" },
      { "src": "./images/02.jpg", "title": "Image 02" },
      
      // More Images Here

      { "src": "./images/20.jpg", "title": "Image 20" },
      { "src": "./images/21.jpg", "title": "Image 21" }
    ]
  }'></div>
```

...**[B]** - Add the JSON object inside the `simbaGrid` function. [View Demo](https://rodgath.github.io/simba-grid/demo/index5.html)

HTML code.
```html
<div class="image-rotation-box image-spin-demo"></div>
```
JavaScript code

```javascript
document.addEventListener('DOMContentLoaded', function() {
   simbaGrid('.image-rotation-box', {
        animationStyle: 'zoom', 
        width: 1300, 
        cols: 6, 
        rows: 4, 
        gap: 10, 
        rowHeight: 200, 
        images: [
          { "src": "./images/01.jpg", "title": "Image 01" },
          { "src": "./images/02.jpg", "title": "Image 02" },
        
        // More Images Here

          { "src": "./images/20.jpg", "title": "Image 20" },
          { "src": "./images/21.jpg", "title": "Image 21" }
        ]
    });
});
```
##### <div align="right"><a href="#top">&uarr; TOP &uarr;</a></div>

---
# Options 

Name | Type | Default | Description
:--- | :--- | :------ | :----------
`width` | _number_ | **1200** | Width of each grid group.
`cols` | _number_ | **3** | Grid columns for each grid group.
`rows` | _number_ | **3** | Grid rows for each grid group.
`rowHeight` | _number_ | **280** | Grid row height.
`gap` | _number_ | **0** | Space between grid items.
`scrollSpeed` | _number_ | **1** | Grid scrolling speed. Min is `1`.
`scrollDirection` | _string_ | **'left'** | Scroll direction. <br> Possible values: `'left'` or `'right'`.
`pauseOnHover` | _boolean_ | **true** | Whether to pause scrolling movement when grid is hovered on.
`shuffle` | _boolean_ | **false** | Whether to shuffle grid items when grid is loaded.
`animationStyle` | _string_ | **'zoom'** | Grid items entry animations. <br> Possible values: `'zoom'`, `'rotate'`, `'zoomRotate'`

## License
simbaGrid is an open-source project released under the [MIT license](http://www.opensource.org/licenses/mit-license.php).

---

Crafted by [Rodgath](https://twitter.com/Rodgath)
##### <div align="right"><a href="#top">&uarr; TOP &uarr;</a></div>

CHANGELOG.MD
###### v1.0.2 - Aug 15, 2023
```
= Update README
```
###### v1.0.1 - Aug 11, 2023
```
= Add support for reverse scrolling
```
###### v1.0.0 - Aug 10, 2023
```
= Initial release.
```

gulpfile.js
const { src, dest, series, parallel, watch, task } = require('gulp');
const uglify = require('gulp-uglify');
const babel = require('gulp-babel');
const rename = require('gulp-rename');
const header = require('gulp-header');
const shell = require('gulp-shell');
const pkg = require('./package.json');

const sourceBranch = 'main';
const targetBranch = 'gh-pages';
const folderToCopy = 'demo';

/* Copy folder between branches */
const copyFolder = () => {
  return shell.task([
    `git checkout ${targetBranch}`,
    `git checkout ${sourceBranch} -- ${folderToCopy}`,
    `git add ${folderToCopy}`,
    `git commit -m "Copy '${folderToCopy}' folder from ${sourceBranch}"`,
    `git push origin ${targetBranch}`,
    `git checkout ${sourceBranch}`
  ]);
}

/* Set banner for dist files */
const setBanner = () => {
	const banner = [
    '/**',
    ` * Simba Grid`,
    ` * @name        ${pkg.name}`,
    ` * @description ${pkg.description}`,
    ` * @link        ${pkg.homepage}`,
    ` * @author      ${pkg.author.name}, ${pkg.author.web}`,
    ` * @version     v${pkg.version}`,
    ` * @created     Jul 22, 2023`,
    ` * @updated     ${pkg.versionDate}`,
    ` * @copyright   Copyright (C) 2023-${new Date().getFullYear()}, ${pkg.author.name}`,
    ` * @license     ${pkg.license}`,
    ` * @licenseMIT  ${pkg.homepage}/blob/main/LICENSE`,
    ` * @demoExample https://rodgath.github.io/simba-grid/demo/`,  
    ' */',
    ''
  ].join('\n');
  
	return src('./dist/{css,js}/**/*.{js,css}')
  .pipe(header(banner))
  .pipe(dest('./dist'))
}

const compressJs = () => {
  return src('./src/*.js')
        .pipe(dest('./demo/js')) // Send original script file to /demo
        .pipe(dest('./dist/js')) // Send original script file to /dist
        .pipe(babel({
            presets: ['@babel/env']
        }))
        .pipe(uglify())
        .pipe(rename( { suffix: '.min' } ))
        .pipe(dest('./demo/js'))
        .pipe(dest('./dist/js'))
}

const buildTask = () => {
  return series(compressJs, setBanner)
}

const watchTask = () => {
  watch(['./src/*.js', './demo/*.html'], { events: 'all' }, buildTask())
}


/* Set tasks */
task('copy:demo', copyFolder());
task('set:banner', setBanner)
task('compress:js', compressJs);
task('watch', watchTask);
task('build', buildTask());

exports.default = buildTask();

src/simba-grid-scroll.js
function simbaGridScroll(element, options) {

  var defaults = {
    width: 1200,
    cols: 3,
    rows: 3,
    rowHeight: 280,
    gap: 0,
    scrollSpeed: 1,
    pauseOnHover: true,
    shuffle: false,
    animationStyle: 'zoom' // 'zoom', 'rotate', 'zoomRotate'
  };
  
  /* Check if options argument is provided directly or in the data attribute */
  if (!options) {
    var dataAttributeOptions = element.dataset.simbaGrid;
    if (dataAttributeOptions) {
      options = JSON.parse(dataAttributeOptions);
    }
  }

  /* Merge 'options' with 'defaults' */
  options = Object.assign({}, defaults, options);

  /* Set image box element */
  let simbaGridWrapper;
  
  /* Check if 'element' is a valid DOM element */
  if (element instanceof HTMLElement || element instanceof Node) {
    simbaGridWrapper = element;
  } else {
    simbaGridWrapper = document.querySelector(element);
  }
  
  const gridItemsObject = getGridItems();
  
  /* Get the images to spin */
  function getGridItems() {

    let clonedWrapper = document.createElement('div');

    let gridItems;
    
    /* Check if ihe 'images' property exists, is an array, and has items */
    if (options.hasOwnProperty('images') && Array.isArray(options.images) && options.images.length > 0) {
      
      options.images.forEach(imageObj => {
        const imgElement = document.createElement('img');
        imgElement.src = imageObj.src;
        imgElement.alt = imageObj.title;
        
        simbaGridWrapper.appendChild(imgElement);
      });

      gridItems = simbaGridWrapper.querySelectorAll('*');
    } else {
      gridItems = simbaGridWrapper.querySelectorAll('*');
    }

    for (let i = 0; i < gridItems.length; i++) {
      clonedWrapper.appendChild(gridItems[i])      
    }
    clonedWrapper.style.display = 'none';
    document.body.appendChild(clonedWrapper);
    // simbaGridWrapper.querySelectorAll('*').remove();
    return Array.from(clonedWrapper.querySelectorAll('*'));
  }
  
  const gridGap = parseInt(options.gap, 10);
  const gridItemHeight = parseInt(options.rowHeight, 10);
  const gridContainerWidth = parseInt(options.width, 10);
  const gridContainerRows = parseInt(options.rows, 10);
  const gridContainerCols = parseInt(options.cols, 10);
  const gridChunkCount = gridContainerRows*gridContainerCols;
  const gridItemsToFetch = gridItemsObject.length+getRandomBetween(1,2);
  const maxGridItemsToFetch = calculateNextDivisible(gridChunkCount, gridItemsToFetch);
  const gridRepeats = maxGridItemsToFetch/gridChunkCount;
  
  const shuffleArray = (array) => {
    
    /* Make a copy of the original array to prevent altering the original array */
    const shuffledArray = array.slice(); // 

    for (let i = shuffledArray.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [shuffledArray[i], shuffledArray[j]] = [shuffledArray[j], shuffledArray[i]];
    }

    return shuffledArray;
  }

  const shuffled = options.shuffle ? shuffleArray(gridItemsObject) : gridItemsObject;
  
  const gridItemsData = repeatArray(shuffled, maxGridItemsToFetch);
  
  
  /* Create a single grid and append images to it */
  const createItemGrid = (itemGroup, num) => {
    
    const simbaGridContainer = document.createElement('div');
    simbaGridContainer.classList.add('simba-grid-container');
    simbaGridContainer.style.minWidth = `${gridContainerWidth}px`;
    
    const simbaGrid = document.createElement('div');
    simbaGrid.classList.add('simba-grid');
    
    simbaGrid.style.display = 'grid';
    simbaGrid.style.gridTemplateColumns = `repeat(${gridContainerCols}, 1fr)`;
    simbaGrid.style.gridAutoRows = `${gridItemHeight}px`;
    simbaGrid.style.gap = `${gridGap}px`;
    
    itemGroup.forEach((itemData, index) => {
      
      const simbaGridItem = document.createElement('div');

      simbaGridItem.classList.add('simba-grid-item');
      // simbaGridItem.style.minHeight = `${gridItemHeight}px`;
      simbaGridItem.style.overflow = 'hidden';
      // simbaGridItem.style.borderRadius = `8px`;
      
      if (itemData.tagName === 'IMG') {
        simbaGridItem.style.backgroundImage = `url(${itemData.currentSrc})`;
        simbaGridItem.style.backgroundSize = 'cover';
        simbaGridItem.style.backgroundPosition = 'center';
      } else {
        itemData.style.width = '100%';
        itemData.style.height = '100%';
        itemData.style.whiteSpace = 'normal';
        simbaGridItem.innerHTML = itemData.outerHTML;
      }
      
      simbaGrid.appendChild(simbaGridItem);
      simbaGridContainer.appendChild(simbaGrid);
    });

    return simbaGridContainer;
  };
  
  /* Get the contents of the JSON object in groups of three */
  const gridGroups = chunkArray(gridItemsData, gridChunkCount);
  const gridGroupsCount = gridGroups.length;
  const gridGroupsCols = gridGroupsCount * 2; // Double the grid for smooth infinite effect
  
  /* Set Simba grid wrapper attributes and styles */
  (function(simbaGridWrapper) {
    simbaGridWrapper.classList.add('simba-grid-wrapper');
    simbaGridWrapper.style.width = '100%';
    simbaGridWrapper.style.whiteSpce = 'nowrap';
    simbaGridWrapper.style.overflow = 'hidden';
    simbaGridWrapper.style.display = 'grid';
    simbaGridWrapper.style.gap = `${gridGap}px`;
    simbaGridWrapper.style.gridTemplateColumns = `repeat(${gridGroupsCols}, 1fr)`;
    simbaGridWrapper.style.gridAutoRows = `${gridItemHeight*gridContainerRows}px`;
    simbaGridWrapper.style.height = `${((gridItemHeight+gridGap)*gridContainerRows)-gridGap}px`;
  })(simbaGridWrapper);
  
  /* Group array elements in sets of 'chunkSize' */
  function chunkArray(arr, chunkSize) {
    const chunkedArray = [];
    for (let i = 0; i < arr.length; i += chunkSize) {
      chunkedArray.push(arr.slice(i, i + chunkSize));
    }
    return chunkedArray;
  }

  gridGroups.forEach((group, index) => {

    const gridContainer = createItemGrid(group, index);
    
    simbaGridWrapper.appendChild(gridContainer);
  })

  cloneAndAppendChildren(simbaGridWrapper);
  
  let isHovering = false;

  /* Handle scrolling */
  const scrollContent = () => {

    const gridContainer = document.querySelector('.simba-grid-container');

    /* Check if the simbaGridWrapper is being hovered */
    if (!isHovering) {
      
      const offset = (gridContainer.offsetWidth * gridRepeats) + (gridRepeats * gridGap);
      
      /* Reset the scroll position to the left when it reaches the end */
      if (simbaGridWrapper.scrollLeft >= offset) {
        simbaGridWrapper.scrollLeft -= offset;
      } else {
        simbaGridWrapper.scrollLeft += options.scrollSpeed;
      }
    }

    /* Request the next animation frame */
    requestAnimationFrame(scrollContent);
  };

  /* Start the scrolling animation */
  scrollContent();

  /* Event listeners to detect hover state */
  simbaGridWrapper.addEventListener('mouseenter', () => isHovering = options.pauseOnHover ? true : false );
  simbaGridWrapper.addEventListener('mouseleave', () => isHovering = false );
  
  function repeatArray(array, targetLength) {

    if (!array.length) return;

    const repeatedArray = [];
    const originalLength = array.length;
    const repeatCount = Math.ceil(targetLength / originalLength);

    for (let i = 0; i < repeatCount; i++) {
      repeatedArray.push(...array);
    }

    return repeatedArray.slice(0, targetLength);
  }
  
  /* Clone and append child elements to the end of the container */
  function cloneAndAppendChildren(container) {
    const originalChildren = container.children;
    const cloneChildren = [];

    /* Clone each child element */
    for (let i = 0; i < originalChildren.length; i++) {
      const originalChild = originalChildren[i];
      const clonedChild = originalChild.cloneNode(true); // Set to true to deep clone with all descendants
      cloneChildren.push(clonedChild);
    }

    /* Append the cloned children to the end of the container */
    cloneChildren.forEach(clonedChild => {
      container.appendChild(clonedChild);
    });

    /* Handle animations */
    const gridItems = container.querySelectorAll('.simba-grid-item')
    for (let j = 0; j < gridItems.length; j++) {
      const gridItem = gridItems[j];
      
      let transformStart = '';
      let transformEnd = '';
      
      switch (options.animationStyle) {
        case 'zoom':
          transformStart = 'scale(0)';
          transformEnd = 'scale(1)';
          break;
        case 'rotate':
          transformStart = 'rotate(-20deg)';
          transformEnd = 'rotate(0deg)';
          break;
        case 'zoomRotate':
          transformStart = 'rotate(-20deg) scale(0)';
          transformEnd = 'rotate(0deg) scale(1)';
          break;
        default:
          transformStart = '';
          transformEnd = '';
          break;
      }
      
      if (options.animationStyle === 'zoom' || 
        options.animationStyle === 'rotate' || 
        options.animationStyle === 'zoomRotate') {
        gridItem.style.opacity = 0;
        gridItem.style['-webkit-transform'] = transformStart;
        gridItem.style.transform = transformStart;
        gridItem.style.transitionProperty = `opacity, transform`;
        gridItem.style.transitionDuration = `0.4s, 0.4s`;
        gridItem.style.transitionTimingFunction = `linear, linear`;
        gridItem.style.transitionDelay = `0s, 0s`;
            
        setTimeout(() => {
          // gridItem.style.transitionDelay = `0.${j*2}s, 0.${j*2}s`;
          gridItem.style.opacity = 1;
          gridItem.style['-webkit-transform'] = transformEnd;
          gridItem.style.transform = transformEnd;
        }, j*200);
      }
    }
  }
  
  function getRandomBetween(min, max) {
    return Math.floor(Math.random() * (max - min + 1) + min);
  }
  
  function calculateNextDivisible(ref, num) {
    
    /* Fix: Ensure grid wrapper has a min of 2 child grids incase Math.ceil(num / ref) is equal to 1 */
    
    /* Calculate the next number that is divisible by reference number */
    const nextDivisible = Math.max(Math.ceil(num / ref), 2) * ref;
    
    return nextDivisible;
  }
  
};

/* Call the function for elements with the 'data-simba-grid' attribute */
const simbaGridScrollElements = document.querySelectorAll('[data-simba-grid]');
simbaGridScrollElements.forEach(element => simbaGridScroll(element));

demo/js/simba-grid.js
function simbaGrid(element, options) {

  var defaults = {
    width: 1200,
    cols: 3,
    rows: 3,
    rowHeight: 280,
    gap: 0,
    scrollSpeed: 1,
    scrollDirection: 'left',
    pauseOnHover: true,
    shuffle: false,
    animationStyle: 'zoom' // 'zoom', 'rotate', 'zoomRotate'
  };
  
  /* Check if options argument is provided directly or in the data attribute */
  if (!options) {
    var dataAttributeOptions = element.dataset.simbaGrid;
    if (dataAttributeOptions) {
      options = JSON.parse(dataAttributeOptions);
    }
  }

  /* Merge 'options' with 'defaults' */
  options = Object.assign({}, defaults, options);

  /* Variable to track scrolling direction */
  let isScrollingRight = options.scrollDirection === 'right' ? true : false;

  /* Set image box element */
  let simbaGridWrapper;
  
  /* Check if 'element' is a valid DOM element */
  if (element instanceof HTMLElement || element instanceof Node) {
    simbaGridWrapper = element;
  } else {
    simbaGridWrapper = document.querySelector(element);
  }
  
  const gridItemsObject = getGridItems();
  
  /* Get the images to spin */
  function getGridItems() {

    let clonedWrapper = document.createElement('div');

    let gridItems;
    
    /* Check if ihe 'images' property exists, is an array, and has items */
    if (options.hasOwnProperty('images') && Array.isArray(options.images) && options.images.length > 0) {
      
      options.images.forEach(imageObj => {
        const imgElement = document.createElement('img');
        imgElement.src = imageObj.src;
        imgElement.alt = imageObj.title;
        
        simbaGridWrapper.appendChild(imgElement);
      });

      gridItems = simbaGridWrapper.querySelectorAll('*');
    } else {
      gridItems = simbaGridWrapper.querySelectorAll('*');
    }

    for (let i = 0; i < gridItems.length; i++) {
      clonedWrapper.appendChild(gridItems[i])      
    }
    clonedWrapper.style.display = 'none';
    document.body.appendChild(clonedWrapper);
    // simbaGridWrapper.querySelectorAll('*').remove();
    return Array.from(clonedWrapper.querySelectorAll('*'));
  }
  
  const gridGap = parseInt(options.gap, 10);
  const gridItemHeight = parseInt(options.rowHeight, 10);
  const gridContainerWidth = parseInt(options.width, 10);
  const gridContainerRows = parseInt(options.rows, 10);
  const gridContainerCols = parseInt(options.cols, 10);
  const gridChunkCount = gridContainerRows*gridContainerCols;
  const gridItemsToFetch = gridItemsObject.length+getRandomBetween(1,2);
  const maxGridItemsToFetch = calculateNextDivisible(gridChunkCount, gridItemsToFetch);
  const gridRepeats = maxGridItemsToFetch/gridChunkCount;
  
  const shuffleArray = (array) => {
    
    /* Make a copy of the original array to prevent altering the original array */
    const shuffledArray = array.slice(); // 

    for (let i = shuffledArray.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [shuffledArray[i], shuffledArray[j]] = [shuffledArray[j], shuffledArray[i]];
    }

    return shuffledArray;
  }

  const shuffled = options.shuffle ? shuffleArray(gridItemsObject) : gridItemsObject;
  
  const gridItemsData = repeatArray(shuffled, maxGridItemsToFetch);
  
  
  /* Create a single grid and append images to it */
  const createItemGrid = (itemGroup, num) => {
    
    const simbaGridContainer = document.createElement('div');
    simbaGridContainer.classList.add('simba-grid-container');
    simbaGridContainer.style.minWidth = `${gridContainerWidth}px`;
    
    const simbaGrid = document.createElement('div');
    simbaGrid.classList.add('simba-grid');
    
    simbaGrid.style.display = 'grid';
    simbaGrid.style.gridTemplateColumns = `repeat(${gridContainerCols}, 1fr)`;
    simbaGrid.style.gridAutoRows = `${gridItemHeight}px`;
    simbaGrid.style.gap = `${gridGap}px`;
    
    itemGroup.forEach((itemData, index) => {
      
      const simbaGridItem = document.createElement('div');

      simbaGridItem.classList.add('simba-grid-item');
      // simbaGridItem.style.minHeight = `${gridItemHeight}px`;
      simbaGridItem.style.overflow = 'hidden';
      // simbaGridItem.style.borderRadius = `8px`;
      
      if (itemData.tagName === 'IMG') {
        simbaGridItem.style.backgroundImage = `url(${itemData.currentSrc})`;
        simbaGridItem.style.backgroundSize = 'cover';
        simbaGridItem.style.backgroundPosition = 'center';
      } else {
        itemData.style.width = '100%';
        itemData.style.height = '100%';
        itemData.style.whiteSpace = 'normal';
        simbaGridItem.innerHTML = itemData.outerHTML;
      }
      
      simbaGrid.appendChild(simbaGridItem);
      simbaGridContainer.appendChild(simbaGrid);
    });

    return simbaGridContainer;
  };
  
  /* Get the contents of the JSON object in groups of three */
  const gridGroups = chunkArray(gridItemsData, gridChunkCount);
  const gridGroupsCount = gridGroups.length;
  const gridGroupsCols = gridGroupsCount * 2; // Double the grid for smooth infinite effect
  
  /* Set Simba grid wrapper attributes and styles */
  (function(simbaGridWrapper) {
    simbaGridWrapper.classList.add('simba-grid-wrapper');
    simbaGridWrapper.style.width = '100%';
    simbaGridWrapper.style.whiteSpce = 'nowrap';
    simbaGridWrapper.style.overflow = 'hidden';
    simbaGridWrapper.style.display = 'grid';
    simbaGridWrapper.style.gap = `${gridGap}px`;
    simbaGridWrapper.style.gridTemplateColumns = `repeat(${gridGroupsCols}, 1fr)`;
    simbaGridWrapper.style.gridAutoRows = `${gridItemHeight*gridContainerRows}px`;
    simbaGridWrapper.style.height = `${((gridItemHeight+gridGap)*gridContainerRows)-gridGap}px`;
  })(simbaGridWrapper);
  
  /* Group array elements in sets of 'chunkSize' */
  function chunkArray(arr, chunkSize) {
    const chunkedArray = [];
    for (let i = 0; i < arr.length; i += chunkSize) {
      chunkedArray.push(arr.slice(i, i + chunkSize));
    }
    return chunkedArray;
  }

  gridGroups.forEach((group, index) => {

    let gridContainer = createItemGrid(group, index);

    if (isScrollingRight) {
      gridContainer = gridItemsReverseOrder(gridContainer);
    }
    
    simbaGridWrapper.appendChild(gridContainer);
  })

  cloneAndAppendChildren(simbaGridWrapper);

  /* Manage grid elements reverse ordering */
  function gridItemsReverseOrder(gridContainer) {
    const grid = gridContainer.querySelector('.simba-grid');
    const squares = grid.querySelectorAll('.simba-grid-item');
    
    // Convert the NodeList to an array for easier manipulation
    const squareArray = Array.from(squares);

    // Reorder the array of squares
    squareArray.sort((a, b) => {
      const aRow = Math.floor(squareArray.indexOf(a) / options.cols); // Divide by 2 since there are 2 columns
      const bRow = Math.floor(squareArray.indexOf(b) / options.cols);
      return bRow - aRow; // Compare row indices to reverse the order
    });

    // Remove existing squares from parent grid
    squares.forEach(square => square.remove());

    // Append the rearranged squares back to the parent grid
    squareArray.forEach(square => grid.appendChild(square));
    
    while (gridContainer.firstChild) {
      gridContainer.removeChild(gridContainer.firstChild);
    }
    
    gridContainer.appendChild(grid)
    const container = gridContainer;
    return container;
  }
  
  let isHovering = false;

  /* Handle scrolling */
  const scrollContent = () => {

    const gridContainer = document.querySelector('.simba-grid-container');

    /* Check if the simbaGridWrapper is being hovered */
    if (!isHovering) {
      
      const offset = (gridContainer.offsetWidth * gridRepeats) + (gridRepeats * gridGap);
      
      if (!isScrollingRight) {
      
        /* Reset the scroll position to the left when it reaches the end */
          if (simbaGridWrapper.scrollLeft >= offset) {
              simbaGridWrapper.scrollLeft -= offset;
          } else {
              simbaGridWrapper.scrollLeft += options.scrollSpeed;
          }
      } else {
          if (simbaGridWrapper.scrollLeft <= offset - gridContainer.parentNode.offsetWidth) { // Check if at the leftmost edge
              simbaGridWrapper.scrollLeft += offset * 2; // Multiply by 2 because the system appends similar number of nodes
          } else {
              simbaGridWrapper.scrollLeft -= options.scrollSpeed;
          }
      }
    }

    /* Request the next animation frame */
    requestAnimationFrame(scrollContent);
  };

  /* Start the scrolling animation */
  scrollContent();

  /* Event listeners to detect hover state */
  simbaGridWrapper.addEventListener('mouseenter', () => isHovering = options.pauseOnHover ? true : false );
  simbaGridWrapper.addEventListener('mouseleave', () => isHovering = false );
  
  function repeatArray(array, targetLength) {

    if (!array.length) return;

    const repeatedArray = [];
    const originalLength = array.length;
    const repeatCount = Math.ceil(targetLength / originalLength);

    for (let i = 0; i < repeatCount; i++) {
      repeatedArray.push(...array);
    }

    return repeatedArray.slice(0, targetLength);
  }

  /* Alter grid items transform animations based on scroll direction */
  function manageGridItemsTransform(gridItems, index) {
    
    const gridItem = gridItems[index];
      
    let transformStart = '';
    let transformEnd = '';

    const rotationAngle = !isScrollingRight ? '-20deg' : '+20deg';
    
    switch (options.animationStyle) {
      case 'zoom':
        transformStart = 'scale(0)';
        transformEnd = 'scale(1)';
        break;
      case 'rotate':
        transformStart = `rotate(${rotationAngle})`;
        transformEnd = 'rotate(0deg)';
        break;
      case 'zoomRotate':
        transformStart = `rotate(${rotationAngle}) scale(0)`;
        transformEnd = 'rotate(0deg) scale(1)';
        break;
      default:
        transformStart = '';
        transformEnd = '';
        break;
    }
    
    if (options.animationStyle === 'zoom' || 
      options.animationStyle === 'rotate' || 
      options.animationStyle === 'zoomRotate') {
      gridItem.style.opacity = 0;
      gridItem.style['-webkit-transform'] = transformStart;
      gridItem.style.transform = transformStart;
      gridItem.style.transitionProperty = `opacity, transform`;
      gridItem.style.transitionDuration = `0.4s, 0.4s`;
      gridItem.style.transitionTimingFunction = `linear, linear`;
      gridItem.style.transitionDelay = `0s, 0s`;
          
      /* Adjust the timing based on reverse/forward direction order */
      const timing = !isScrollingRight ? index * 200 : (gridItems.length - index - 1) * 200;
      
      setTimeout(() => {
        // gridItem.style.transitionDelay = `0.${index*2}s, 0.${index*2}s`;
        gridItem.style.opacity = 1;
        gridItem.style['-webkit-transform'] = transformEnd;
        gridItem.style.transform = transformEnd;
      }, timing);
    }
  }
  
  /* Clone and append child elements to the end of the container */
  function cloneAndAppendChildren(container) {
    const originalChildren = container.children;
    const cloneChildren = [];

    /* Clone each child element */
    for (let i = 0; i < originalChildren.length; i++) {
      const originalChild = originalChildren[i];
      const clonedChild = originalChild.cloneNode(true); // Set to true to deep clone with all descendants
      cloneChildren.push(clonedChild);
    }

    /* Append the cloned children to the end of the container */
    cloneChildren.forEach(clonedChild => {
      container.appendChild(clonedChild);
    });

    /* Handle animations */
    const gridItems = container.querySelectorAll('.simba-grid-item')
    if (!isScrollingRight) {
      for (let j = 0; j < gridItems.length; j++) {
        manageGridItemsTransform(gridItems, j);
      }
    } else {
      for (let j = gridItems.length - 1; j >= 0; j--) { // Loop in reverse order
        manageGridItemsTransform(gridItems, j);
      }
    }
    
    if (isScrollingRight) {
      for (let i = 0; i < originalChildren.length; i++) {
        const originalChild = originalChildren[i];
        gridItemsReverseOrder(originalChild)
      }
    }
  }
  
  function getRandomBetween(min, max) {
    return Math.floor(Math.random() * (max - min + 1) + min);
  }
  
  function calculateNextDivisible(ref, num) {
    
    /* Fix: Ensure grid wrapper has a min of 2 child grids incase Math.ceil(num / ref) is equal to 1 */
    
    /* Calculate the next number that is divisible by reference number */
    const nextDivisible = Math.max(Math.ceil(num / ref), 2) * ref;
    
    return nextDivisible;
  }
  
};

/* Call the function for elements with the 'data-simba-grid' attribute */
const simbaGridScrollElements = document.querySelectorAll('[data-simba-grid]');
simbaGridScrollElements.forEach(element => simbaGrid(element));